### PR TITLE
OpenACC + Cray CCE + AMD MI200+

### LINKED ISSUE #352: [bug] Rebuilding HDF5
We appear to (fully) rebuild HDF5 with every instantiation of `./mfc.sh run [...] -t post_process` (at least on GPU). It takes non-negligible time (at least a few minutes).

* Computer: Phoenix
* Mode: Interactive

### PR DESCRIPTION
**Pull request by anandrdbz**
## Description

Adds support for MI200+ GPUs via CCE compilers and OpenACC.

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

### Scope

- [x] This PR comprises a set of related changes with a common goal

Closes #352 #383 #384 

**Test Configuration**:

* What computers and compilers did you use to test this:
	* OLCF Frontier

### CODE REVIEW

**Review by sbryngelson — CHANGES_REQUESTED** [review]
file `toolchain/run/:wq` is a vim mistake

**Comment by sbryngelson** [comment]
Conflicts need to be resolved

File: misc/gnu-acc.sh
Comment (sbryngelson) [inline misc/gnu-acc.sh:1]: what is this file, and why does it show up as new?

File: toolchain/mfc/run/mpi_bins.py
Comment (sbryngelson) [inline toolchain/mfc/run/mpi_bins.py:1]: is this an actual new file? it seems like an old file...

**Comment by sbryngelson** [comment]
Aside from the test suite, the benchmarks are also failing on GPU:
```
Case                     Pre Process   Simulation   Post Process  
 ────────────────────────────────────────────────────────────────── 
  viscous_weno5_sgb_mono         1.00x          N/A            N/A  
  5eq_rk3_weno3_hllc             0.50x        0.98x          1.33x  
  ibm                            1.00x        1.04x          1.00x  
  hypo_hll                       1.00x          N/A            N/A
```

File: examples/3D_weak_scaling/case.py [hunk 1]
```diff
@@ -186,8 +186,8 @@
     'cyl_coord'                    : 'F',
     'dt'                           : dt,
     't_step_start'                 : 0,
-    't_step_stop'                  : int(5000*16.0/ARGS["gbpp"]),
-    't_step_save'                  : int(1000*16.0/ARGS["gbpp"]),
+    't_step_stop'                  : int(25),
+    't_step_save'                  : int(25),
```
Comment (henryleberre) [inline examples/3D_weak_scaling/case.py]: Should these changes be reverted?

Comment (sbryngelson) [reply examples/3D_weak_scaling/case.py]: yes

File: profile.md
Comment (sbryngelson) [inline profile.md:1]: please move this file into `/misc/` and rename it `profiling_amdgpu.md`

File: examples/3D_testcase/case_pre_sim.py
Comment (sbryngelson) [inline examples/3D_testcase/case_pre_sim.py:1]: what's this file? delete?

File: src/common/m_variables_conversion.fpp [hunk 2]
```diff
@@ -1242,9 +1280,19 @@ contains
                             + pres_K*dir_flg(dir_idx(i))
                     end do
 
+
+                    if(j == 1) then
+#ifdef CRAY_PRINT_DEBUG
```
Comment (sbryngelson) [inline src/common/m_variables_conversion.fpp]: delete these lines?

File: src/pre_process/m_start_up.fpp [hunk 3]
```diff
@@ -837,7 +837,7 @@ contains
 &                pTg-equilirium (relax = "T" activated)'
             end if
 
-            call s_relaxation_solver(q_cons_vf)
+            call s_infinite_relaxation_k(q_cons_vf)
```
Comment (sbryngelson) [inline src/pre_process/m_start_up.fpp:840]: is this intentional? seems wrong?

File: src/simulation/m_compute_cbc.fpp [hunk 4]
```diff
@@ -28,7 +28,11 @@ contains
         !!      the normal component of velocity is zero at all times,
         !!      while the transverse velocities may be nonzero.
     subroutine s_compute_slip_wall_L(lambda, L, rho, c, mf, dalpha_rho_ds, dpres_ds, dvel_ds, dadv_ds) ! --------------
+#ifdef CRAY_ACC_WAR
+        !DIR$ INLINEALWAYS s_compute_slip_wall_L
```
Comment (sbryngelson) [inline src/simulation/m_compute_cbc.fpp:32]: is it ok that this statement is indented? I'm not sure. May need to ask @abbotts

Comment (sbryngelson) [reply src/simulation/m_compute_cbc.fpp:32]: nevermind it is fine https://docs.oracle.com/cd/E19957-01/805-4941/6j4m2sobl/index.html

File: src/simulation/m_global_parameters.fpp [hunk 5]
```diff
@@ -20,7 +20,7 @@ module m_global_parameters
 
     use m_derived_types        !< Definitions of the derived types
 
-#ifdef MFC_OpenACC
+#ifdef MFC_OPENACC
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp]: @anandrdbz this is a bug (should be `#ifdef MFC_OpenACC`)

File: src/simulation/m_global_parameters.fpp [hunk 6]
```diff
@@ -310,23 +355,43 @@ module m_global_parameters
...
+    !$acc declare create( k_n, k_v, pb0, mass_n0, mass_v0, Pe_T, Re_trans_T, Re_trans_c, Im_trans_T, Im_trans_c, omegaN)
+#endif
     real(kind(0d0)) :: mul0, ss, gamma_v, mu_v
     real(kind(0d0)) :: gamma_m, gamma_n, mu_n
     real(kind(0d0)) :: gam
     !> @}
-    !$acc declare create(R_n, R_v, phi_vn, phi_nv, Pe_c, Tw, pv, M_n, M_v, k_n, k_v, pb0, mass_n0, mass_v0, Pe_T, Re_trans_T, Re_trans_c, Im_trans_T, Im_trans_c, omegaN , mul0, ss, gamma_v, mu_v, gamma_m, gamma_n, mu_n, gam)
+
+    !$acc declare create(mul0, ss, gamma_v, mu_v, gamma_m, gamma_n, mu_n, gam)
+
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp]: @anandrdbz i think this might be where some bugs live. you have gone from a long list of variables in declare create to splitting it between a shorter list and then, in the `#ifdef CRAY_ACC_WAR` statement on line 411, we have declare global/link for the others. however, if we use NVHPC then those additional variables are never declared!

Comment (sbryngelson) [reply src/simulation/m_global_parameters.fpp]: actually, it looks like they are declared.

File: src/simulation/m_global_parameters.fpp [hunk 7]
```diff
@@ -612,7 +692,7 @@ contains
                     ! print*, 'alf idx', alf_idx
                     ! print*, 'bub -idx beg end', bub_idx%beg, bub_idx%end
 
-                    @:ALLOCATE(weight(nb), R0(nb), V0(nb))
+                    @:ALLOCATE_GLOBAL(weight(nb), R0(nb), V0(nb))
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp:689]: is this correct for nvhpc?

File: src/simulation/m_global_parameters.fpp [hunk 8]
```diff
@@ -891,21 +971,33 @@ contains
         intxe = internalEnergies_idx%end
 
         !$acc update device(momxb, momxe, advxb, advxe, contxb, contxe, bubxb, bubxe, intxb, intxe, sys_size, buff_size, E_idx, alf_idx, strxb, strxe)
+        !$acc update device(m, n, p)
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp:967]: is this right for nvhpc?

File: src/simulation/m_global_parameters.fpp [hunk 9]
```diff
@@ -891,21 +971,33 @@ contains
...
 
-        ! Allocating grid variables for the x-, y- and z-directions
-        @:ALLOCATE(x_cb(-1 - buff_size:m + buff_size))
-        @:ALLOCATE(x_cc(-buff_size:m + buff_size))
-        @:ALLOCATE(dx(-buff_size:m + buff_size))
+        !$acc update device(alt_soundspeed, monopole, num_mono)
+        !$acc update device(dt, sys_size, buff_size, pref, rhoref, gamma_idx, pi_inf_idx, E_idx, alf_idx, stress_idx, mpp_lim, bubbles, hypoelasticity, alt_soundspeed, avg_state, num_fluids, model_eqns, num_dims, mixture_err, grid_geometry, cyl_coord, mapped_weno, mp_weno, weno_eps)
```
Comment (sbryngelson) [inline src/simulation/m_global_parameters.fpp:970]: is this new code right for nvhpc?

File: src/simulation/m_global_parameters.fpp (hunk 7 above)
Comment (abbotts) [reply src/simulation/m_global_parameters.fpp:689]: For nvhpc (and CCE 17, once I verify we can turn off CRAY_ACC_WAR) `ALLOCATE_GLOBAL` and `ALLOCATE` should be identical. I'll skim the standard to see if there's reason you may want to differentiate between module allocatables and locally scoped allocatables in the future, besides the general fragility of FORTRAN module variables in directive languages.

File: src/simulation/m_global_parameters.fpp (hunk 6 above)
Comment (anandrdbz) [reply src/simulation/m_global_parameters.fpp]: These are deliberate changes for CRAY, all the dynamic arrays need to be declared separately for CRAY v/s NVHPC (also if this was the case you would have bugs in 1D too).

File: src/simulation/m_global_parameters.fpp (hunk 7 above)
Comment (sbryngelson) [reply src/simulation/m_global_parameters.fpp:689]: good to know. there is a bug causing this code to not run on nvhpc + Nvidia GPUs at the moment, so I was checking out the changes to see what might be an issue.

Comment (anandrdbz) [reply src/simulation/m_global_parameters.fpp:689]: Yes, I have rewritten all the macros so that it does the same thing as before

File: src/simulation/m_global_parameters.fpp (hunk 8 above)
Comment (anandrdbz) [reply src/simulation/m_global_parameters.fpp:967]: Hmm I don't see how this could be an issue, there might be duplicate updates but that shouldn't make a difference anyways

File: src/simulation/m_global_parameters.fpp (hunk 9 above)
Comment (anandrdbz) [reply src/simulation/m_global_parameters.fpp:970]: These statements are just moved around I think, and allocate changed to allocate_global (which does the same thing)

File: src/simulation/m_global_parameters.fpp (hunk 5 above)
Comment (anandrdbz) [reply src/simulation/m_global_parameters.fpp]: Yeah this is a bug, will fix it, but it's inconsequential

File: src/pre_process/m_start_up.fpp (hunk 3 above)
Comment (anandrdbz) [reply src/pre_process/m_start_up.fpp:840]: It's intentional, function pointers don't work for phase change

File: src/common/m_variables_conversion.fpp (hunk 2 above)
Comment (anandrdbz) [reply src/common/m_variables_conversion.fpp]: will do

**Review by sbryngelson — APPROVED** [review]

File: src/common/m_variables_conversion.fpp (hunk 2 above)
Comment (sbryngelson) [reply src/common/m_variables_conversion.fpp]: reminder

File: examples/3D_weak_scaling/case.py (hunk 1 above)
Comment (sbryngelson) [reply examples/3D_weak_scaling/case.py]: @anandrdbz revert these changes please

Comment (anandrdbz) [reply examples/3D_weak_scaling/case.py]: done

File: examples/3D_weak_scaling/case.py [hunk 10]
```diff
@@ -199,7 +199,7 @@
     'mpp_lim'                      : 'T',
     'mixture_err'                  : 'T',
     'time_stepper'                 : 3,
-    'weno_order'                   : 3,
+    'weno_order'                   : 5,
```
Comment (sbryngelson) [inline examples/3D_weak_scaling/case.py]: why the change?

File: src/common/m_phase_change.fpp [hunk 11]
```diff
@@ -74,7 +74,7 @@ contains
         !!      selecting the phase change module that will be used
         !!      (pT- or pTg-equilibrium)
     subroutine s_initialize_phasechange_module()
-
+!#ifndef _CRAYFTN
```
Comment (sbryngelson) [inline src/common/m_phase_change.fpp]: ? what is this

File: src/common/m_variables_conversion.fpp [hunk 12]
```diff
@@ -1244,6 +1276,7 @@ contains
 
                     ! energy flux, u(E+p)
                     FK_vf(j, k, l, E_idx) = vel_K(dir_idx(1))*(E_K + pres_K)
+                    ! comment out above and it will run to completion
```
Comment (sbryngelson) [inline src/common/m_variables_conversion.fpp]: ?

File: src/common/m_variables_conversion.fpp [hunk 13]
```diff
@@ -1244,6 +1276,7 @@ contains
...
 
                     ! have been using == 2
```
Comment (sbryngelson) [inline src/common/m_variables_conversion.fpp]: what is this? remove?

File: src/simulation/m_rhs.fpp [hunk 14]
```diff
@@ -1,4 +1,4 @@
-!>
+!!>
```
Comment (sbryngelson) [inline src/simulation/m_rhs.fpp]: why the change?

File: examples/3D_weak_scaling/case.py (hunk 10 above)
Comment (anandrdbz) [reply examples/3D_weak_scaling/case.py]: Ben had done it to examine profiles for WENO5, I will revert back to original

**Comment by sbryngelson** [comment]
Builing on a Frontier login noce, `salloc`'ing into a compute node, then running test gives:
```
scratch/MFC $ ./mfc.sh test -- -c frontier
mfc: OK > (venv) Entered the Python 3.10.10 virtual environment (>= 3.8).

Traceback (most recent call last):
  File "/ccs/home/sbryngelson/scratch/MFC/toolchain/main.py", line 5, in <module>
    from mfc         import args, lock, build, bench, state, count
  File "/lustre/orion/cfd154/scratch/sbryngelson/MFC/toolchain/mfc/args.py", line 3, in <module>
    from .run.run      import get_baked_templates
  File "/lustre/orion/cfd154/scratch/sbryngelson/MFC/toolchain/mfc/run/run.py", line 5, in <module>
    from mako.lookup   import TemplateLookup
  File "/ccs/home/sbryngelson/scratch/MFC/build/venv/lib/python3.10/site-packages/mako/lookup.py", line 15, in <module>
    from mako.template import Template
  File "/ccs/home/sbryngelson/scratch/MFC/build/venv/lib/python3.10/site-packages/mako/template.py", line 20, in <module>
    from mako import codegen
  File "/ccs/home/sbryngelson/scratch/MFC/build/venv/lib/python3.10/site-packages/mako/codegen.py", line 16, in <module>
    from mako import filters
  File "/ccs/home/sbryngelson/scratch/MFC/build/venv/lib/python3.10/site-packages/mako/filters.py", line 14, in <module>
    import markupsafe
ModuleNotFoundError: No module named 'markupsafe'

mfc: ERROR > mfc.py finished with a 1 exit code.
mfc: (venv) Exiting the Python virtual environment.
```

**Comment by sbryngelson** [comment]
Change `./mfc.sh load` compute name from Crusher to Frontier

**Comment by sbryngelson** [comment]
`./mfc.sh test -a -- -c frontier` does not work

**Comment by sbryngelson** [comment]
It's looking like Frontier CI may fail for the 2-rank case. Tests were run with
```
./mfc.sh test -j 8 -- -c frontier
```

The test MFC.sh file in the 2-rank directory reads
```
        (set -x;                 srun -N 1                                    -n         2                                                                     "/lustre/orion/    cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/build/install/0571538fd2/ bin/simulation")
```

which appears to be the problem, it should be passing `--ntasks-per-node` (or whatever) since we are using `-- -c frontier`

File: examples/3D_weak_scaling/case.py (hunk 10 above)
Comment (wilfonba) [reply examples/3D_weak_scaling/case.py]: This change was made because the performance results on NVIDIA hardware we were comparing to use WENO5 and this is the case we were using for performance results on AMD hardware.

**Comment by sbryngelson** [comment]
@henryleberre, do you know why it doesn't build h5dump? (or at least it isn't found in the expected `bin/` directory)

**Comment by henryleberre** [comment]
@sbryngelson We opted not to build HDF5 on CCE. I forget why, perhaps there were some incompatibilities. We use the cray-hdf5 module so hdf5dump should already be available.

**Comment by sbryngelson** [comment]
@henryleberre you are correct, `h5dump` is already in the path. It looks like the problem is that using `test -a` forces it to look in `dependencies/bin/h5dump` for the binary (rather than the path broadly). Is there is a fix for this?

**Comment by henryleberre** [comment]
@sbryngelson I'm testing a fix. For your command, you would have use this instead:
```console
$ ./mfc.sh test -a --no-hdf5 -- -c frontier
```

**Comment by sbryngelson** [comment]
@henryleberre this works!

**Comment by sbryngelson** [comment]
closes #352 #383 #384
